### PR TITLE
Completely define external repositories

### PR DESCRIPTION
**Pull request by jart**
This PR adds:

- Redundant mirror URLs to all downloads
- Total, strict, minimal, and customizable definition of Java build graph
- Simplification of web_test_repositories function because of how downloads work
- Introduction of java_import_external which Nomulus, Closure Rules, etc. already use

To learn more about the best practices used in this CL see: https://github.com/bazelbuild/bazel/issues/1952, https://github.com/bazelbuild/bazel/commit/ed7ced0018dc5c5ebd6fc8afc7158037ac1df00d, and ping me on chat for more info.

### CODE REVIEW

**Comment by DrMarcII** [comment]
I would prefer that SeleniumServer have the standalone jar pulled rather than the individual maven jars. In practice it is being used purely as a binary launched by web_test when needed. The only reason I created a java_binary rule was to ensure that it got an appropriate jvm to run with (there is probably a better way of doing this, but I haven't quite figured out what that is).

Additionally, SeleniumServer more appropriately belongs in browser_repositories rules.

**Comment by jart** [comment]
If we used the standalone jar for selenium server, we're talking like *three* fewer java_import_external rules. Nearly all these java_import_external rules come from org.seleniumhq.selenium:selenium-java:3.0.1. So why pay the price of downloading the code twice?

> The only reason I created a java_binary rule was to ensure that it got an appropriate jvm to run with (there is probably a better way of doing this, but I haven't quite figured out what that is).

I'm not sure I understand what you mean about appropriate jvm. I will say this though. I consider strictly linear dependencies to be an objective good. It takes more effort, but it avoids less evil in the grand scheme of things. Life is better for everyone when we work harder.

**Comment by DrMarcII** [comment]
Please rebase and update.

**Comment by jart** [comment]
Done. PTAL

File: web/repositories.bzl [hunk 1]
```diff
@@ -13,86 +13,95 @@
...
+  with the omit parameters. This is useful for users who want to be rigorous
+  about declaring their own direct dependencies, or when another Bazel project
+  is depended upon (e.g. rules_closure) that defines the same dependencies as
+  this one (e.g. com_google_guava.) Alternatively, a whitelist model may be
+  used by calling the individual functions this method references.
+
+  Please note that while these dependencies are defined, they are not actually
+  downloaded, unless a target is built that depends on them.
   """
+  _check_bazel_version("Webtesting Rules", "0.4.2")
```
Comment (DrMarcII) [inline web/repositories.bzl]: Web Testing Rules?

File: web/repositories.bzl [hunk 2]
```diff
@@ -105,37 +114,381 @@ def browser_repositories(firefox=False, chromium=False):
     firefox: Configure repositories for //browsers:firefox-native.
     chromium: Configure repositories for //browsers:chromium-native.
   """
+  _check_bazel_version("Webtesting Rules", "0.4.2")
```
Comment (DrMarcII) [inline web/repositories.bzl]: Web Testing Rules?

**Comment by jart** [comment]
Happy to address both of those suggestions. PTAL

**Pull request merged**